### PR TITLE
Various improvements to Quantum Volume

### PR DESCRIPTION
**Pull request by KevinVillela**
#2111 @mpharrigan 
- Make TwoQubitMatrixGate and QuantumVolumeResult JSON serializable/loadable
- Make a number of routing attempts and choose the best one (configurable by a flag - defaults to 30).
  - Also adds a Jupyter notebook analyzing the impact of this change:
image
- Refactor the algorithm to first compile all the circuits, and then run them all on one sampler at a time.

### CODE REVIEW

File: cirq/contrib/quantum_volume/quantum_volume.py [hunk 1]
```diff
@@ -165,21 +167,30 @@ def compile_circuit(
...
     if router is None and routing_algo_name is None:
         routing_algo_name = 'greedy'
-    swap_network = ccr.route_circuit(compiled_circuit,
-                                     ccr.xmon_device_to_graph(device),
-                                     router=router,
-                                     algo_name=routing_algo_name)
-    compiled_circuit = swap_network.circuit
+
+    best_swap_network: Union[ccr.SwapNetwork, None] = None
+    best_score = -1
```
Comment (mpharrigan) [inline cirq/contrib/quantum_volume/quantum_volume.py]: extreme nit: this seems like a weird initial value. I typically use `None` for these cases.

File: cirq/protocols/json.py [hunk 2]
```diff
@@ -97,6 +99,7 @@ def cirq_class_resolver_dictionary(self) -> Dict[str, Type]:
                 'pandas.DataFrame': pd.DataFrame,
                 'pandas.Index': pd.Index,
                 'pandas.MultiIndex': pd.MultiIndex.from_tuples,
+                'TwoQubitMatrixGate': cirq.TwoQubitMatrixGate,
```
Comment (mpharrigan) [inline cirq/protocols/json.py]: heads up: pr #2381 seems to be deprecated soon cc @Strilanc

**Comment by mpharrigan** [comment]
@Strilanc how do you want to handle the forthcoming merge conflict with #2381

File: cirq/contrib/quantum_volume/quantum_volume.py (hunk 1 above)
Comment (KevinVillela) [reply cirq/contrib/quantum_volume/quantum_volume.py]: Done.

**Comment by KevinVillela** [comment]
> @Strilanc how do you want to handle the forthcoming merge conflict with #2381

Ah, didn't see that. I'm happy to wait until that PR is in and then merge, unless you think it will take a while?

**Comment by KevinVillela** [comment]
Thanks @Strilanc, much appreciated!

**Pull request merged**